t(emporal)PARAFAC2's temporal smoothness penalty

cchatzis opened the pull request:
Implementation of temporal smoothness penalty, as presented in:

> C. Chatzis, M. Pfeffer, P. Lind and E. Acar, "A Time-Aware Tensor Decomposition for Tracking Evolving Patterns," 2023 IEEE 33rd International Workshop on Machine Learning for Signal Processing (MLSP), Rome, Italy, 2023, pp. 1-6, doi: 10.1109/MLSP55844.2023.10285943.

This PR essentially adds the custom penalty class, (`TemporalSmoothnessPenalty`) to `penalties.py`.

MarieRoald commented on src/matcouply/penalties.py:
```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
+    @copy_ancestor_docstring
+    def factor_matrices_update(self, factor_matrices, feasibility_penalties, auxes):
+
+        B_is = factor_matrices
```
```suggestion
        B_is = factor_matrices
        I = len(factor_matrices)  # noqa: E741
```

It might be more readible if we add a variable for the number of matrices.

MarieRoald commented on src/matcouply/penalties.py:
```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
+        B_is = factor_matrices
+        rhos = feasibility_penalties
+
+        rhs = [rhos[i] * factor_matrices[i] for i in range(len(B_is))]
```
```suggestion
        rhs = [rhos[i] * factor_matrices[i] for i in range(I)]
```

MarieRoald commented on src/matcouply/penalties.py:
```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
+
+        # Construct matrix A to peform thomas algorithm on
+
+        A = np.zeros((len(B_is) , len(B_is)))
+
+        for i in range(len(B_is)):
+            for j in range(len(B_is)):
+                if i == j:
+                    A[i , j] = 4 * self.smoothness_l + rhos[i]
+                elif (i == j - 1) or (i == j + 1):
+                    A[i , j] = - 2 * self.smoothness_l
+                else:
+                    pass
+
+        A[0, 0] -= 2 * self.smoothness_l
+        A[len(B_is) - 1 , len(B_is) - 1] -= 2 * self.smoothness_l
```
```suggestion
     A = (
        tl.diag([self._get_laplace_coef(i) + rho for i, rho in enumerate(rhos)], k=0)
        - tl.diag(tl.ones(I-1) * 2 * self.smoothness_l, k=1)
        - tl.diag(tl.ones(I-1) * 2 * self.smoothness_l, k=-1)
    )
```

This will reduce the assembly complexity from O(I^2) to O(I). We'd need the `_get_laplace_coef` method though, which could look like this for example:

```python
def _get_laplace_coef(self, i, feasibility_penalties):
    I = len(feasibility_penalties)
    rhos = feasibility_penalties
    if i == 0 or i = I - 1:
        return 2 * self.smoothness_l + rhos[i]
    return 4 * self.smoothness_l + rhos[i]
```

cchatzis replied:
This was addressed in 95877352ccbba80b65ed9eaa865d55f68814dd68.

MarieRoald commented on src/matcouply/penalties.py:
```diff
@@ -1322,3 +1322,98 @@ def penalty(self, x):
+
+        # Back-substitution
+
+        new_ZBks = [np.empty_like(B_is[i]) for i in range(len(B_is))]
```
```suggestion
        new_ZBks = [np.empty_like(B_is[i]) for i in range(I)]
```

MarieRoald reviewed (changes requested):
Hi @cchatzis,

Thank you for the PR! This looks like a great contribution -- a perfect fit for MatCoupLy! 

I made a couple of small suggestions, including one that avoids a nested for loop for the A-assembly, which could speed it up for large `I`.

It would also be great with a test for this penalty. To add this, you need to add a test class that inherits from `matcouply.testing.BaseTestFactorMatricesPenalty` (it should already be imported in `tests/test_penalties.py`) and add three methods: `test_penalty`, `get_invariant_matrices` and `get_non_invariant_matrices`:

 * `test_penalty` should test that the penalty is computed correctly, either based on a hard-coded example where you know what the penalty should be, or by computing it completely manually in the test
 * `get_invariant_matrices(rng, shapes)` should generate a random list of matrices that won't be changed by the proximal operator (I guess it should be a stack of the same matrix repeated `len(shapes)` times?).
 * `get_non_invariant_matrices(rng, shapes)` should generate a random list of matrices that will be changed by the proximal operator (I guess it could just be a list of random matrices).

It might be useful to set the class attribute `min_rows = max_rows = 10` or something to ensure that all shapes in the `shapes` list are the same. The tests for the `GeneralizedL2Penalty` might also be a useful place to start: https://github.com/MarieRoald/matcouply/blob/main/tests/test_penalties.py#L247

Don't be afraid to ask if you have any other questions or if anything I said wasn't clear :)

cchatzis commented:
Thank you very much for the thorough suggestions and help @MarieRoald!

I think I have addressed all of the raised issues and I added a small test `test_A_assembly` in `tests/test_penalties.py` that checks that both the efficient approach you proposed and "hard-coded" approach I had initially have the same result. 

Kindly let me know what you think!

MarieRoald commented on tests/test_penalties.py:
```diff
@@ -759,3 +763,44 @@ def test_penalty(self, random_ragged_cmf):
+        for k in range(1, K):
+            penalty_term += tl.norm(B_is[k] - B_is[k - 1]) ** 2
+
+        assert np.isclose(penalty.penalty(B_is), penalty_term)
```
I prefer using the `assert_allclose` function rather than asserting the output of `np.isclose` because it works with any TensorLy tensor

```suggestion
        assert_allclose(penalty.penalty(B_is), penalty_term)
```

MarieRoald commented on tests/test_penalties.py:
```diff
@@ -759,3 +763,44 @@ def test_penalty(self, random_ragged_cmf):
+
+        penalty_term = 0
+        for k in range(1, K):
+            penalty_term += tl.norm(B_is[k] - B_is[k - 1]) ** 2
```
This is a minor nit, but if we use `tl.sum` instead of doing the norm, then we'll get fewer roundoff errors and it might be slightly (but probably not measurably) faster as we skip a square root followed by a square

```suggestion
            penalty_term += tl.sum((B_is[k] - B_is[k - 1]) ** 2)
```

MarieRoald commented on tests/test_penalties.py:
```diff
@@ -759,3 +763,44 @@ def test_penalty(self, random_ragged_cmf):
+        K = len(shapes)
+
+        invariant_matrix = np.array(rng.random_sample(shapes[0]))
+        invariant_matrices = [np.copy(invariant_matrix) for _ in range(K)]
```
We need to wrap this in `tl.tensor` to be sure that it works for PyTorch too

```suggestion
        invariant_matrices = [tl.tensor(invariant_matrix.copy()) for _ in range(K)]
```

MarieRoald commented on tests/test_penalties.py:
```diff
@@ -759,3 +763,44 @@ def test_penalty(self, random_ragged_cmf):
+
+        K = len(shapes)
+
+        non_invariant_matrices = [rng.random_sample(shapes[k]) for k in range(K)]
```
Same as above

```suggestion
        non_invariant_matrices = [tl.tensor(rng.random_sample(shapes[k])) for k in range(K)]
```

MarieRoald commented on src/matcouply/penalties.py:
```diff
@@ -1394,12 +1394,10 @@ def factor_matrices_update(self, factor_matrices, feasibility_penalties, auxes):
-            np.diag([self._get_laplace_coef(i, K, self.smoothness_l) + rhos[i] for i, rho in enumerate(rhos)], k=0)
-            - np.diag(np.ones(K - 1) * 2 * self.smoothness_l, k=1)
-            - np.diag(np.ones(K - 1) * 2 * self.smoothness_l, k=-1)
+            np.diag([self._get_laplace_coef(i, I) + rhos[i] for i, rho in enumerate(rhos)], k=0)
+            - np.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=1)
+            - np.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=-1)
```
If we use `tl.diag`, we'll support PyTorch as well :)

```suggestion
            tl.diag([self._get_laplace_coef(i, I) + rhos[i] for i, rho in enumerate(rhos)], k=0)
            - tl.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=1)
            - tl.diag(np.ones(I - 1) * 2 * self.smoothness_l, k=-1)
```

MarieRoald reviewed (changes requested):
Looks really good, just a couple of minor comments. The PyTorch are failing, but addressing the comments will hopefully make them pass :)

cchatzis commented:
Hi, are there any updates on this? Thanks in advance :)

MarieRoald commented:
Looks good, but we need to know if the tests run. For some reason, the CI/CD pipeline doesn't want to trigger, so I made some changes to the `.github/workflows/Tests.yml` on main.

I tried to push the changes directly on top of your fork, but I wasn't able to. So I think that either, you need to merge main into your feature branch or make it so I can push to your branch.


To merge, I think it should work if you:

 1. Make sure that my repo is added as a remote. Run `git remote -v` to list all remotes, there should be a remote with the name `upstream` (or something similar) that points to my repo. If there is no remote with the url `https://github.com/MarieRoald/matcouply.git` (or the SSH-equivalent), then you should run `git remote add upstream https://github.com/MarieRoald/matcouply.git`.
 2. Check out main and pull latest changes: `git checkout main` followed by `git pull upstream`.
 3. Finally, check out your branch and merge main: `git checkout tPARAFAC2` followed by `git merge main`.

cchatzis commented:
Hi, 

I followed the steps. Let me know if I have done something incorrectly.

Thanks in advance!
